src/nvim/autocmd.c

```diff
@@ -925,6 +925,18 @@ static int do_autocmd_event(event_T event, char_u *pat, bool once, int nested, c
             return FAIL;
           }
         }
+
+        // need to initialize last_mode for the first ModeChanged autocmd
+        if (event == EVENT_MODECHANGED && !has_event(EVENT_MODECHANGED)) {
+          char *mode = get_mode();
+          if (last_mode == NULL) {
+            last_mode = (char *)vim_strsave((char_u*)mode);
+          } else {
+            STRCPY(last_mode, mode);
+          }
+          xfree(mode);
```

Comment:
`get_mode()` is an allocated string.
What's the issue with freeing `last_mode` first and then assigning the new value to it?

Reply:
In the original Vim patch `last_mode` is an array. If `last_mode` is changed to an array here then this will make sense.

Reply:
See https://github.com/neovim/neovim/pull/15840#discussion_r740912878 for why I use a pointer instead of an array.

I agree with @janlazo that just freeing and assigning is better. The if branch comes from a misunderstanding of mine, which was that I thought freeing NULL would be a problem (which it obviously isn't).
I have implemented the proposed change now.